### ISSUE TITLE
deprecate nix-vscode-extensions in favor of nix4vscode

### ISSUE THREAD

**Issue by deemp** [nix-community/nix-vscode-extensions#47]
This issue is a result of the discussion of #41.

## nix-vscode-extensions

Provides the latest versions of almost all extensions and release versions of chosen extensions from VS Code Marketplace and Open VSX.

### Advantages

- flake provides extension patches for all users ([link](https://github.com/nix-community/nix-vscode-extensions#special-extensions))
- doesn't require configuration files
- supports VS Code Marketplace and Open VSX
- flake is updated automatically via a GH Action
- provides platform-specific extensions
- extension info is in JSON format and can be reused without nix

### Disadvantages

- flake is large due to files with extensions config
- doesn't provide extension versions compatible with the user's VS Code
- not very maintainable due to a) Haskell and b) non-professional code
- possibly can't be scaled to support release versions for all VS Code versions ([link](https://github.com/nix-community/nix-vscode-extensions/issues/41#issuecomment-1709103703))
- latest versions include pre-release versions that expire soon

## nix4vscode

[nix4vscode](https://github.com/cathaysia/nix4vscode) provides the latest versions compatible with a given VS Code version (see #41) for chosen extensions.

### Advantages

- provides extension versions compatible with the user's VS Code
- generates expressions only for necessary extensions
- written in Rust and seems to be maintainable
- can be possibly scaled to all extensions and used in `nixpkgs` ([link](https://github.com/nix-community/nix-vscode-extensions/issues/41#issuecomment-1694480585))

### Disadvantages

- doesn't provide extension patches for all users
- requires a configuration file
- a user has to generate expressions
- doesn't provide platform-specific extensions (`extensionFromVscodeMarketplace` doesn't use `platform`)
- extension info is in Nix and cannot be easily reused without `nix`
- doesn't support Open VSX
- is currently unstable

## Summary

Despite some disadvantages, the approach of `nix4vscode` seems to be more promising than that of `nix-vscode-extensions`. In my opinion, most of `nix4vscode` disadvantages can be overcome. Then, it will be a truly personalized tool for getting VS Code extensions.

## Suggestions

@deemp, @AmeerTaweel: 
- Don't implement the solution ([link](https://github.com/nix-community/nix-vscode-extensions/issues/41#issuecomment-1687764105))
- Decide if need to deprecate `nix-vscode-extensions` and at what stage of readiness of `nix4vscode`
- If decide to eventually deprecate:
  - Edit README
    - Warn in the README of `nix-vscode-extensions` about deprecation and link this issue
    - Put into README a link to the `nix4vscode` repo
  - Edit wiki ([link](https://nixos.wiki/wiki/Visual_Studio_Code#NixOS)):
     - Link to `nix4vscode` instead of `nix-vscode-extensions`

@cathaysia
- Transfer `nix4vscode` to `nix-community` to get more contributors
  - Perhaps @spikespaz will be interested to work on `nix4vscode` as it's written in Rust
- Consider listed disadvantages and turn some of them into issues in the `nix4vscode` repo

**Comment by spikespaz** [comment]
Thanks for ping.

**Comment by cathaysia** [comment]
Thanks, I will forward nix4vscode to nix-community and list the issue you mentioned in the issue. Thank you very much for your trust in nix4vscode. As of now, my configuration uses both nix-vscode-extensions and nix4vscode, if you plan to deprecate nix-vscode-extensions, I would try to completely transfer the configuration to nix4vscode for further testing.

**Comment by deemp** [comment]
`nix-vscode-extensions` will stay here at least until `nix4vscode` is mostly feature-complete and stable.

**Comment by AmeerTaweel** [comment]
- `nix4vscode` is added to `nix-community`.
- The last commit was two weeks ago.

@cathaysia is `nix4vscode` mostly feature-complete and stable?

**Comment by cathaysia** [comment]
No, there's still a lot of work to be done, but a simple scenario (generating nix expressions from a configuration file) should be enough, and the likelihood of changes is unlikely. If everything goes well, I will complete these tasks before the Spring Festival this year.
________________________________

**Comment by spikespaz** [comment]
@cathaysia 

I have not had the time in months to look at the project, but I want to see a way to use it entirely from Nix without IFD. Not sure how that would look, but I don't really want to have to run a separate command to update everything. I want to write expressions that choose versions based on Nix config.

Until that happens, nix4vscode cannot replace this repo for me.

**Comment by spikespaz** [comment]
Additionally, I elected myself for that job. I was thinking of generating a shell script (from the rust binary) that sets up environment and links, which can be used in the activation scripts or shell nix.

**Comment by cathaysia** [comment]
Work is in progress to provide something like nix-vscode-extensions. But you know, I need to work and rest, and I also need time to fall in love, so I don’t have much time left for open source work. The functions required by nix4vscode are not complicated, and one to two weeks should be enough to write and test, but this requires a whole block of time, rather than always squeezing out some scattered time. In addition, if you have good ideas to propose in nix4vscode, PRs are always welcome.

**Comment by spikespaz** [comment]
I wasn't trying to rush you. I too have not been able to work on it, because I do not live exclusively in a code editor (contrary to what some may believe).

**Comment by cathaysia** [comment]
thank you for understanding. :)

**Comment by cathaysia** [comment]
Hello, I recently implemented a similar function in nix4vscode. 

- https://github.com/nix-community/nix4vscode/blob/25d1f0e2796a1af08243a9d8684e507a8632e339/flake.nix#L74  
- https://github.com/nix-community/nix4vscode/blob/25d1f0e2796a1af08243a9d8684e507a8632e339/example/flake.nix#L32

After comparing toml and json, I found that toml is slightly smaller than json after serialization. So I chose toml as the archive format.

However, the number of exported extensions is much less than I expected, and many extensions do not appear in the exported toml file. Do you know how to query all the extension information in vscode-marketplace?

**Comment by cathaysia** [comment]
ci will periodically save all information into a sqlite3 database, which you can get here: 
https://github.com/nix-community/nix4vscode/tree/db

**Comment by AmeerTaweel** [comment]
Sorry for the late reply @cathaysia

It's been a long while since I worked on this repo's code. As a matter of fact, I don't understand most of it since @deemp rewrote it in Haskell. I learned some Haskell since then but I only know the basics of the basics. On the other hand, my understanding of Rust, while not deep, is much better than my understanding of Haskell, as I use it to solve competitive programming questions.

I looked at @deemp's implementation and your implementation for getting the list of extensions to try and spot the difference. I think I caught something interesting, but I'm not sure if I'm right or not.

What I Spotted:

@deemp's implementation retries failed API queries (`getConfigs` and `getConfigsRelease` in `haskell/app/Main.hs`), while your code doesn't (`get_extension_response` in `crates/code_api/src/code/http_client.rs`).

So I think that your code stops sending API requests on the first failure, and if failures happen often enough, you end up with a small list of extensions. While @deemp's code retries the failed queries so it gets much more extensions.

Please try changing this and see if it fixes the issue.

Again, sorry for the late reply.

**Comment by AmeerTaweel** [comment]
Unrelated to your question, I was reading Farid Zakaria's blog series on Nix Dynamic Derivations:

1. [An Early Look at Nix Dynamic Derivations](https://fzakaria.com/2025/03/10/an-early-look-at-nix-dynamic-derivations)
2. [Nix Dynamic Derivations: A Practical Application](https://fzakaria.com/2025/03/11/nix-dynamic-derivations-a-practical-application)
3. [Nix Dynamic Derivations: A Lang2Nix Practicum](https://fzakaria.com/2025/03/12/nix-dynamic-derivations-a-lang2nix-practicum)

And I was wondering if we could benefit from this feature in `nix-vscode-extensions` and `nix4vscode`. Farid's examples don't map cleanly to the VSCode extensions use-case but it feels very similar.

**Comment by deemp** [comment]
I'll close this issue because I don't plan to deprecate the project given that many people use it.

**Comment by cathaysia** [comment]
Hi, @deemp have you paid attention to the changes in nix4vscode recently? I think nix4vscode is very good now.
